Update settings menu

louh opened the pull request:
Resolves #1006.

louh commented:
@treyhahn FYI -- This adds new strings for the "units" settings.

mkong0216 commented:
Not sure if I'm the only one seeing this, but as seen by the screenshot below, my username seems to be written twice and over each other?

![screen shot 2018-06-08 at 11 47 34 am](https://user-images.githubusercontent.com/17012261/41168651-9430e622-6b14-11e8-9a0e-5ed05ea769a7.png)

Also for the settings dropdown, is it possible to make the dropdown look more clickable and maybe have more padding?

louh commented:
It's _supposed_ to look like this:

<img width="1245" alt="screenshot 2018-06-08 12 14 09" src="https://user-images.githubusercontent.com/2553268/41169028-a80d6ac0-6b15-11e8-8d79-9a5796824d29.png">

Try doing a hard refresh? (We should probably look into a cache-busting URL of some kind for the CSS...)

treyhahn commented:
I have 2 thoughts on this:
1. Do we still want to maintain this other place to change the units?

![image](https://user-images.githubusercontent.com/12552044/41183249-b1b5f6e6-6b47-11e8-9a0a-b5ede71126a9.png)

2. What do people think about changing "settings" to just the language title like @mwichary mentioned here: https://github.com/streetmix/streetmix/issues/1006#issue-327934012? Or, since units are now included in the menu too would it be better not to do that?

louh commented:
> Do we still want to maintain this other place to change the units?

For now, yes. There may be users who will be shocked if it disappears suddenly. Ideally, we would monitor how people change units and wean people off the existing method over time.

> What do people think about changing "settings" to just the language title like @mwichary mentioned here: #1006 (comment)? Or, since units are now included in the menu too would it not make sense to do that?

This was my thought.

treyhahn commented:
Thanks for the responses. Agreed, that makes sense.

ryross commented on assets/scripts/menus/LocaleDropdown.jsx:
```diff
@@ -1,109 +1,68 @@
-  onChange = (event) => {
-    this.props.changeLocale(event.target.value)
+    locale: DEFAULT_LOCALE,
+    level: 4,
```
Why do we default this to 4? Can we have have a comment here? Do other components pass in level1, level2 or level3 props that need to be changed as well?

louh replied:
The different languages are classified as different levels of readiness, with "4" being "this is the default language" (English). Level 3 is any other "ready" language, level 2 is "close but testing" and level 1 is "not ready at all; just getting started".

I left a markdown file describing these here: https://raw.githubusercontent.com/streetmix/streetmix/master/app/data/locales.md

This is not obvious so I'll definitely leave a comment or figure out some other way of making this clearer.

ryross commented on assets/scripts/menus/LocaleDropdown.jsx:
```diff
@@ -1,109 +1,68 @@
+    return filteredLocales.map((locale) => {
+      const classNames = ['menu-item']
+
+      if (locale.value === selectedLocale) {
```
Would it make more sense to do the selected locale check here in the map instead of having another loop above?

louh replied:
I was trying to figure out a one-loop way to do this:

1. Given a list of filtered locales (which can change depending on other criteria, see other note about levels)
2. and given a selected locale of any value (passed in as props)
3. assume that it is possible for a selected locale to not be present in the list of filtered locales
4. therefore it must default to another locale that is presently in the list (this is `DEFAULT_LOCALE`, or `"en"`, which is always in the list)

Is there a way to do this without searching through the entire filtered list (the first loop) first?

If there were only one loop, and it found a match later in the list, but English already matched first, how do you go back and un-match English?

ryross commented on assets/scripts/menus/SettingsMenu.jsx:
```diff
@@ -1,17 +1,92 @@
+function mapStateToProps (state) {
+  // The lowest level marked "true" takes priority.
+  let level = 4
+  if (state.flags.LOCALES_LEVEL_3.value) level = 3
```
How do these locale levels work? Are certain levels hidden by default or something? Might be beneficial to have a comment here relaying that.

louh replied:
Right, so level 4 is always on (this is English, the language we've always had).
Levels 1-3 were off by default. We flipped level 3 (ready to go, non English) to "on" to make those languages available to everyone. Levels 1 and 2 will not be turned permanently on but we envision we would flip those switches for certain users/environments for testing.

If this system doesn't make any sense, let me know if you think we can improve it.

louh commented:
@ryross I moved the "language verification" step out of the `LocaleSelect` component (which I also renamed, now that it's not strictly a dropdown). This is good because the logic can be used elsewhere; there were a few other locale-related bugs in existence previously:

- If a user momentarily had access to "level 1" languages and selected (for instance, German), then that access was later revoked, the user would still continue to be able to use German until they switched to another language.
- If a user had a native browser language that wasn't supported by Streetmix, Streetmix would attempt to use it instead of falling back to English, causing a ton of `react-intl` error messages to appear in the console.
- This includes "subset" languages like `en-US`, which previously had a hack to force it to `en`, but it didn't help if you were in another country, like Australia. This "subset" mismatch was likely to have been the cause of the settings menu occasionally displaying languages [in brackets]. (cc @treyhahn)


Another adjustment was made to make sure that locale loading completed during init, so that one would not load Streetmix in English first only to have it converted to another language split-seconds later.

louh commented:
Another UI issue: it's possible for the language to load while an InfoBubble is open, which causes it to re-render (as an entirely new component), losing its original visual position. The re-render occurs on the top left (as in, zero left and zero top). Moving the mouse to re-hover on a segment fixes this, but the initial effect will be jarring.

louh commented:
I've added a loading state to the locale selection menu. Now, if loading takes a while (it's fetching data across a network) we have a little feedback as to why there will be a small delay between the time the menu closes and the languages update the app.

This keeps the menu open, and effectively prevents an InfoBubble from displaying during the loading process (after which it will be re-rendered and its position reset, which is what we wanted to avoid).

Note: since the app causes re-mounting of the entire React tree, it also means the menu doesn't actually ever "know" that the language has changed because it's destroyed and remounted once it has. I've had to add some code in the `<MenuBar />` component to address this -- it also fixes a current problem, which is that when the languages change and the app remounts, a "ghost" menu is still active, which blocks UI interaction until someone clicks the screen.
